### PR TITLE
Add linting of fixtures SQL for critical rules errors to tox

### PR DESCRIPTION
**Pull request by tunetheweb**
### Brief summary of the change made

We have a wealth of example SQL files, and we test them for parsing in various environments. However we don't lint them.

We have had some recent issues (see #2464, #2467, #2469) whereby certain SQL caused rule code to throw a critical error, mostly because the rule hasn't been built to consider certain scenarios.

This PR adds a script to lint all the `test/fixtures/dialects` files, but only complain if a CRITICAL error is thrown by SQLFluff (normal SQLFluff linting errors are thrown aside as most test cases are not written with all the rules in mind - and shouldn't be IMHO).

It's a simple bash script (it's what I know best!) and changes to tox and our CI GitHub Action. We could change it to a python script (in line with the rest of the code), but this'll do for now. Pus I can of like the fact it's calling SQLFluff by command line as it would. I'll leave one of the more experienced Python developers to take a conversion to Python on if they want!

Note this cannot be merged until #2464, #2468, #2470 are all merged as currently it (correctly!) highlights errors, that those were raised to fix.

### Are there any other side effects of this change that we should be aware of?
I also added a .sqlfluff config file to each dialect directory to automatically set the dialect for that directory. This is automatically taken care of in our test suite, but think it'll be handy anyway for when running sqlfluff outside the test suite (like we do here).

At the moment our Windows CI takes the longest to run, so will time this, but suspect (based on local testing on my slow Mac) it'll be quicker than that so shouldn't add to CI run times.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by tunetheweb** [comment]
Confirmed it takes 4 mins to run, so won't hold up CI.

**Review by pwildenhain — COMMENTED** [review]
Looks like a great addition

File: test/test-lint-rules.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+DIALECTS=($( ls -d test/fixtures/dialects/*/ ))
+for DIALECT in "${DIALECTS[@]}"
+do
+    echo "Testing $DIALECT SQL files lint without critical errors..."
+    OUTPUT=$(sqlfluff lint ${DIALECT})
```
Comment (barrywhart) [inline test/test-lint-rules.sh:7]: Should we (or do we already) check the `sqlfluff` exit code here as well?

**Review by barrywhart — APPROVED** [review]
So beautiful, I'm almost in tears. Very nice addition!!

One small question/suggestion, otherwise 💯.

**Comment by barrywhart** [comment]
@tunetheweb: It looks like this is not currently a "required" CI check. Can you make it required, or does @alanmcruickshank have to do that?

File: test/test-lint-rules.sh (hunk 1 above)
Comment (tunetheweb) [reply test/test-lint-rules.sh:7]: We can't. There are LOTS of linting errors (fixtures aren't written to follow all the rules). So it mostly exists with exit code 65. But that's not to say it must. A new dialect could be written with one perfect test case which will exit with 0.

So let's leave as unchecked.

**Comment by tunetheweb** [comment]
> @tunetheweb: It looks like this is not currently a "required" CI check. Can you make it required, or does @alanmcruickshank have to do that?

Only @alanmcruickshank can do that. Let's discuss in this afternoon's meeting.

**Comment by tunetheweb** [comment]
> Let's discuss in this afternoon's meeting.

Sorry next week's meeting :-)

**Review by barrywhart — APPROVED** [review]

**Comment by NiallRees** [comment]
Would definitely be good to get this in Python so that it can be run by Windows users, but not a blocker!

**Pull request merged**